### PR TITLE
Introduce a test for fuzzing engines

### PR DESCRIPTION
**Pull request by ligurio**

### CODE REVIEW

**Comment by ligurio** [comment]
blocked by:
```
2024-07-17T13:18:44.7260913Z tarantool: ./src/box/memtx_engine.cc:1258: void memtx_engine_abort_checkpoint(engine*): Assertion `memtx->checkpoint != NULL' failed.
```

File: test/fuzz/lua/test_engine.lua [hunk 1]
```diff
@@ -0,0 +1,1411 @@
...
+        field_count = oneof({0, table.getn(space_format)}),
+        format = space_format,
+        if_not_exists = oneof({true, false}),
+        is_local = oneof({true, false}),
+    }
+    if space_opts.engine ~= 'vinyl' then
+        space_opts.temporary = oneof({true, false})
+    end
+    local space_name = ('test_%d'):format(space_id_func())
+    local space = box.schema.space.create(space_name, space_opts)
```
Comment (locker) [inline test/fuzz/lua/test_engine.lua:536]: Let's create a few test spaces (say 3-4), not just one.

File: test/fuzz/lua/test_engine.lua [hunk 2]
```diff
@@ -0,0 +1,1411 @@
...
+    index_create_op(space)
+    log.info('FINISH SETUP')
+    return space
+end
+
+local function teardown(space, dir)
+    log.info('TEARDOWN')
+    space:drop()
+    if dir ~= nil then
+        rmtree(dir)
```
Comment (locker) [inline test/fuzz/lua/test_engine.lua:547]: Removing files from under Tarantool may result in unpredictable behavior. Let's leave the directory be. Also, let's add an option for the work dir (current directory by default).

File: test/fuzz/lua/test_engine.lua [hunk 3]
```diff
@@ -0,0 +1,1411 @@
...
+-- Forward declaration.
+local index_create_op
+
+local function random_int()
+    return math.floor(math.random() * 10^12)
+end
+
+local RTREE_DIMENSION = oneof({2, 4, 8})
+
+local function random_array()
```
Comment (Gumix) [inline test/fuzz/lua/test_engine.lua:185]: Why is this called `random_array()`? There are no `math.random()` like in `random_map()` bellow.

File: test/fuzz/lua/test_engine.lua (hunk 2 above)
Comment (ligurio) [reply test/fuzz/lua/test_engine.lua:547]: With a patch below by default, test uses directory `test_engine_dir` in a current directory. Custom test directory can be specified with option `--test_dir`. The script will clean up the directory before testing if it exists.

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index 995ef2d3c9..aba1724edd 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -19,11 +19,15 @@ local datetime = require('datetime')
 local decimal = require('decimal')
 local uuid = require('uuid')
 
+local test_dir_name = 'test_engine_dir'
+local DEFAULT_TEST_DIR = fio.pathjoin(fio.cwd(), test_dir_name)
+
 local params = require('internal.argparse').parse(arg, {
     { 'engine', 'string' },
     { 'h', 'boolean' },
     { 'seed', 'number' },
     { 'test_duration', 'number' },
+    { 'test_dir', 'string' },
     { 'verbose', 'boolean' },
     { 'workers', 'number' },
 })
@@ -51,20 +55,21 @@ end
 local index_id_func = counter()
 
 if params.help or params.h then
-    print([[
+    print(([[
 
  Usage: tarantool test_engine.lua [options]
 
  Options can be used with '--', followed by the value if it's not
  a boolean option. The options list with default values:
 
-   workers <number, 50>          - number of fibers to run simultaneously
-   test_duration <number, 2*60>  - test duration time (sec)
-   engine <string, 'vinyl'>      - engine ('vinyl', 'memtx')
-   seed <number>                 - set a PRNG seed
-   verbose <boolean, false>      - enable verbose logging
-   help (same as -h)             - print this message
-]])
+   workers <number, 50>                  - number of fibers to run simultaneously
+   test_duration <number, 2*60>          - test duration time (sec)
+   test_dir <string, ./%s>  - path to a test directory
+   engine <string, 'vinyl'>              - engine ('vinyl', 'memtx')
+   seed <number>                         - set a PRNG seed
+   verbose <boolean, false>              - enable verbose logging
+   help (same as -h)                     - print this message
+]]):format(test_dir_name))
     os.exit(0)
 end
 
@@ -74,6 +79,9 @@ local arg_num_workers = params.workers or 50
 -- Test duration time.
 local arg_test_duration = params.test_duration or 2*60
 
+-- Test directory.
+local arg_test_dir = params.test_dir or DEFAULT_TEST_DIR
+
 -- Tarantool engine.
 local arg_engine = params.engine or 'vinyl'
 
@@ -125,15 +133,15 @@ local function keys(t)
     return table_keys
 end
 
-local function rmtree(s)
-    log.info(('CLEANUP %s'):format(s))
-    if (fio.path.is_file(s) or fio.path.is_link(s)) then
-        fio.unlink(s)
+local function rmtree(path)
+    log.info(('CLEANUP %s'):format(path))
+    if (fio.path.is_file(path) or fio.path.is_link(path)) then
+        fio.unlink(path)
         return
     end
-    if fio.path.is_dir(s) then
-        for _, i in pairs(fio.listdir(s)) do
-            rmtree(s..'/'..i)
+    if fio.path.is_dir(path) then
+        for _, p in pairs(fio.listdir(path)) do
+            rmtree(fio.pathjoin(path, p))
         end
     end
 end
@@ -543,15 +551,19 @@ local function setup(engine_name, space_id_func, test_dir, verbose)
     return space
 end
 
-local function teardown(space, dir)
-    log.info('TEARDOWN')
-    space:drop()
+local function cleanup_dir(dir)
+    log.info('CLEANUP')
     if dir ~= nil then
         rmtree(dir)
         dir = nil -- luacheck: ignore
     end
 end
 
+local function teardown(space)
+    log.info('TEARDOWN')
+    space:drop()
+end
+
 -- Indexes,
 -- https://www.tarantool.io/en/doc/latest/concepts/data_model/indexes/
 -- space_object:create_index(),
@@ -1358,11 +1370,17 @@ local function process_errors(error_messages)
     return found_unexpected_errors
 end
 
-local function run_test(num_workers, test_duration, engine_name, verbose_mode)
-    local workers = {}
+local function run_test(num_workers, test_duration, test_dir,
+                        engine_name, verbose_mode)
+
+    if fio.path.exists(test_dir) then
+        cleanup_dir(test_dir)
+    else
+        fio.mkdir(test_dir)
+    end
 
+    local workers = {}
     local space_id_func = counter()
-    local test_dir = fio.tempdir()
     local space = setup(engine_name, space_id_func, test_dir, verbose_mode)
 
     local test_gen = fun.cycle(fun.iter(keys(ops)))
@@ -1405,10 +1423,11 @@ local function run_test(num_workers, test_duration, engine_name, verbose_mode)
         end
     end
 
-    teardown(space, test_dir)
+    teardown(space)
 
     local exit_code = process_errors(error_messages) and 1 or 0
     os.exit(exit_code)
 end
 
-run_test(arg_num_workers, arg_test_duration, arg_engine, arg_verbose)
+run_test(arg_num_workers, arg_test_duration, arg_test_dir,
+         arg_engine, arg_verbose)
```

File: test/fuzz/lua/test_engine.lua (hunk 1 above)
Comment (ligurio) [reply test/fuzz/lua/test_engine.lua:536]: I would postpone this and make the test script more complicated when current version will stop to find new bugs. Sure, there is a room for improvements in the test script, but we should put it to a master as soon as possible to make it available for everyone.

File: test/fuzz/lua/test_engine.lua (hunk 3 above)
Comment (ligurio) [reply test/fuzz/lua/test_engine.lua:185]: Actually `random_array` uses random dimension `RTREE_DIMENSION` that *generated* only once at start, see above.

I've updated the function a bit:

```diff
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -181,10 +181,16 @@ local function random_int()
     return math.floor(math.random() * 10^12)
 end
 
-local RTREE_DIMENSION = oneof({2, 4, 8})
+-- Maximal possible R-tree dimension,
+-- see <src/lib/salad/rtree.h>.
+local RTREE_MAX_DIMENSION = 20
 
+local RTREE_DIMENSION = math.random(RTREE_MAX_DIMENSION)
+
+-- RTREE is a single index that support arrays, length of arrays
+-- depends on a RTREE's dimension.
 local function random_array()
-    local n = RTREE_DIMENSION
+    local n = RTREE_DIMENSION * 2
     local arr = {}
     for i = 1, n do
         table.insert(arr, i)
```

**Review by locker — APPROVED** [review]

**Review by Gumix — APPROVED** [review]

**Comment by ligurio** [comment]
There was a problem when error injections (`ERRINJ_WAL_FALLOCATE` and `ERRINJ_BUILD_INDEX`) were not applied due to wrong arguments. The patch below fixes arguments in err injections (index name was replaced with index id), fixes comments for err injections and adds  `errinj_f` fiber object to a table with workers. All errors reported by errinj fiber will be printed at the end of the test.

The patch below:

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index e1c5bbac9b..f193ce4c38 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -1068,7 +1068,7 @@ local errinj_set = {
     -- alter.
     ERRINJ_BUILD_INDEX = {
         enable = function(space)
-            return oneof(keys(space.index))
+            return math.random(#(keys(space.index)))
         end,
         disable = function(_space)
             return -1
@@ -1176,7 +1176,7 @@ local errinj_set = {
         enable = enable_errinj_boolean,
         disable = disable_errinj_boolean,
     },
-    -- Set to true to fail vinyl index dump.
+    -- Set to timeout to fail vinyl index dump.
     ERRINJ_VY_INDEX_DUMP = {
         enable = enable_errinj_timeout,
         disable = function() return -1 end,
@@ -1262,7 +1262,7 @@ local errinj_set = {
         enable = enable_errinj_timeout,
         disable = disable_errinj_timeout,
     },
-    -- Set to true to fail allocation of vinyl tuple.
+    -- Set to timeout to fail allocation of vinyl tuple.
     ERRINJ_VY_STMT_ALLOC = {
         enable = enable_errinj_timeout,
         disable = disable_errinj_timeout,
@@ -1283,11 +1283,11 @@ local errinj_set = {
         enable = enable_errinj_boolean,
         disable = disable_errinj_boolean,
     },
-    -- Set to true to fail WAL write due to error allocating disk
+    -- Set to timeout to fail WAL write due to error allocating disk
     -- space.
     ERRINJ_WAL_FALLOCATE = {
-        enable = enable_errinj_boolean,
-        disable = disable_errinj_boolean,
+        enable = enable_errinj_timeout,
+        disable = disable_errinj_timeout,
     },
     -- Set to true to fail WAL write.
     ERRINJ_WAL_IO = {
@@ -1319,7 +1319,7 @@ local errinj_set = {
         enable = enable_errinj_boolean,
         disable = disable_errinj_boolean,
     },
-    -- Set to true to fail write to xlog file.
+    -- Set to timeout to fail write to xlog file.
     ERRINJ_WAL_WRITE_PARTIAL = {
         enable = enable_errinj_timeout,
         disable = disable_errinj_timeout,
@@ -1329,7 +1329,7 @@ local errinj_set = {
         enable = enable_errinj_boolean,
         disable = disable_errinj_boolean,
     },
-    -- Set to true to fail xlog data read.
+    -- Set to timeout to fail xlog data read.
     ERRINJ_XLOG_READ = {
         enable = enable_errinj_timeout,
         disable = disable_errinj_timeout,
@@ -1402,7 +1402,7 @@ local function run_test(num_workers, test_duration, test_dir,
     end, arg_test_duration)
     errinj_f:set_joinable(true)
     errinj_f:name('ERRINJ')
-    errinj_f:join()
+    table.insert(workers, errinj_f)
 
     local error_messages = {}
     for _, fb in ipairs(workers) do
```

**Review by ylobankov — APPROVED** [review]

**Pull request merged**